### ISSUE TITLE
Unexpected Exception Thrown when Opening Inaccessible/Non-existent Spreadsheet using 'open_by_url' Method

### ISSUE THREAD

**Issue by amey-neve** [burnash/gspread#1244]
**Description**
I encountered an unexpected behavior while working with the gspread (5.9.0) library, specifically when attempting to open an inaccessible or non-existent spreadsheet using the `open_by_url()` method. I expected the library to raise a `SpreadsheetNotFound` error in such cases, as it does when using the `open()` method. However, instead of the expected exception, the library throws an `APIError`.

**To Reproduce**
1. Import the gspread library and the necessary dependencies.
2. Use the open_by_url method with the URL of an inaccessible or non-existent spreadsheet.
3. Observe the exception thrown by the library.

**Expected behavior**
I expected the open_by_url method to raise a SpreadsheetNotFound exception when attempting to open an inaccessible or non-existent spreadsheet.

Actual Behavior:
Contrary to expectations, the open_by_url method throws an APIError when provided with the URL of an inaccessible or non-existent spreadsheet.

**Code example***
```
>>> gc = gspread.service_account_from_dict(service_account_key)
>>> sh = gc.open_by_url("https://docs.google.com/spreadsheets/d/non-existent-url/edit")
>>> sh.worksheets()
gspread.exceptions.APIError: {'code': 404, 'message': 'Requested entity was not found.', 'status': 'NOT_FOUND'}
```

**Environment info:**
- Operating System: MacOS
- Python version: 3.11
- gspread version: 5.9.0

**Comment by lavigne958** [comment]
Hi I understand your issue.

In fact when using the method `open_by_url` or `open_by_key`, we don't actually _open_ the file.

gspread creates a new object of tyep `Spreadsheet` and simply gets the metadata for that spreadsheet.

When the file does not exists we don't do anything.

We can add some coverage around that:
- when opening the spreadsheet file get the metadata
- if we fail with: "404 entity was not found"
- we can raise instead `SpreadsheetNotFound` which is cleaner

We'll make a new release this week, I am not too sure we have time to fit in before, if not it will be coming in the next release after this week.

**Comment by amey-neve** [comment]
Thank you for the prompt response. Looking forward to the upcoming release. 👍 

One more thing, I did not find any custom exception for `Permission Denied` response.
I think one more exception can be added to handle it. Something like, `SpreadsheetPermissionDenied` which is more cleaner.

**Comment by lavigne958** [comment]
you're welcome.

no we don't we can add one too matching specifically this exception too. that's a good idea.

**Comment by alifeee** [comment]
This sounds like a good change to me. Thus, I implemented it in #1245.

In my opinion, it is a breaking change, as people may rely on the current Exceptions, so I have added it to the `6.0.0` release.

Thanks for the issue!
